src/primitives/compression.h

```diff
@@ -0,0 +1,475 @@
+        return a.m_txid == b.m_txid && a.m_n == b.m_n;
+    }
+    friend bool operator!=(const CCompressedOutPoint& a, const CCompressedOutPoint& b) {
+        return !(a == b);
+    }
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        this->m_txid.Serialize(s);
+        s << VARINT(this->m_n);
+    }
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s, bool compressedTxId, bool constructor=false) {
+        if (!constructor) this->m_txid.Unserialize(s, compressedTxId);
+        this->m_n = std::numeric_limits<uint32_t>::max();
+        s >> VARINT(this->m_n);
+    }
+
+    template <typename Stream>
+    explicit CCompressedOutPoint(deserialize_type, Stream& s, bool compressedTxId) : m_txid(CCompressedTxId(deserialize, s, compressedTxId)) {
+        Unserialize(s, compressedTxId, true);
+    }
+
+    std::string ToString() const;
+};
+
+class CCompressedTxIn
+{
+private:
+    CCompressedOutPoint m_prevout;
+    std::vector<unsigned char> m_signature;
+    bool m_compressedSignature;
+    uint8_t m_hashType;
+    uint32_t m_nSequence;
+public:
+    const CCompressedOutPoint& prevout() const { return m_prevout; }
+    const std::vector<unsigned char>& signature() const { return m_signature; }
+    const bool& compressedSignature() const { return m_compressedSignature; }
+    const uint8_t& hashType() const { return m_hashType; }
+    const uint32_t& nSequence() const { return m_nSequence; }
+
+    explicit CCompressedTxIn(secp256k1_context* ctx, const CTxIn& txin, const CCompressedTxId& txid, const CScript& scriptPubKey);
+
+    bool IsSigned() const {
+        return this->m_signature.size() > 0;
+    }
+
+    bool IsSequenceStandard() const {
+        return this->m_nSequence == 0x00 || this->m_nSequence == 0xFFFFFFF0 || this->m_nSequence == 0xFFFFFFFE || this->m_nSequence == 0xFFFFFFFF;
+    }
+
+    bool IsHashStandard() const {
+        return this->m_hashType == 0x00 || this->m_hashType == 0x01;
+    }
+
+    friend bool operator==(const CCompressedTxIn& a, const CCompressedTxIn& b) {
+        return a.m_prevout == b.m_prevout && a.m_signature == b.m_signature && a.m_compressedSignature == b.m_compressedSignature && a.m_hashType == b.m_hashType && a.m_nSequence == b.m_nSequence;
+    }
+
+    friend bool operator!=(const CCompressedTxIn& a, const CCompressedTxIn& b) {
+        return !(a == b);
+    }
+
+    uint64_t GetMetadata() const {
+        uint64_t metadata = 0;
+        metadata |= this->IsSigned();
+        metadata |= this->m_compressedSignature << 1;
+        metadata |= this->m_prevout.txid().IsCompressed() << 2;
+        metadata |= this->IsHashStandard() << 3;
+        metadata |= (this->m_hashType > 0) << 4;
+        metadata |= this->IsSequenceStandard() << 5;
+        metadata |= ((this->m_nSequence == 0xFFFFFFF0) ? 0x01 : (this->m_nSequence == 0xFFFFFFFE) ? 0x02 : (this->m_nSequence == 0xFFFFFFFF) ? 0x03 : 0x00) << 6;
```

Comment:
In 89b52782f8d6aca75d49048da1e137fbf615c0e6:

I'd reformat this as a `switch` statement. (There is a way to format long chains of `?:`s, but this isn't it, and regardless, it's bad style.)

Reply:
Solved, Thanks!